autoprecompiles/src/memory_optimizer.rs

```diff
@@ -4,62 +4,59 @@ use std::hash::Hash;
 
 use itertools::Itertools;
 use powdr_constraint_solver::boolean_extractor::{self, RangeConstraintsForBooleans};
-use powdr_constraint_solver::constraint_system::{ConstraintRef, ConstraintSystem};
+use powdr_constraint_solver::constraint_system::{BusInteraction, ConstraintRef, ConstraintSystem};
 use powdr_constraint_solver::indexed_constraint_system::IndexedConstraintSystem;
-use powdr_constraint_solver::quadratic_symbolic_expression::RangeConstraintProvider;
 use powdr_constraint_solver::quadratic_symbolic_expression::{
-    NoRangeConstraints, QuadraticSymbolicExpression,
+    QuadraticSymbolicExpression, RangeConstraintProvider,
 };
 use powdr_constraint_solver::utils::possible_concrete_values;
-use powdr_number::{FieldElement, LargeInt};
-
-use crate::legacy_expression::{AlgebraicExpression, AlgebraicReference};
-use crate::optimizer::algebraic_to_quadratic_symbolic_expression;
-use crate::{SymbolicBusInteraction, SymbolicConstraint, SymbolicMachine};
+use powdr_number::FieldElement;
 
 /// The memory address space for register memory operations.
 const REGISTER_ADDRESS_SPACE: u32 = 1;
 
 /// Optimizes bus sends that correspond to general-purpose memory read and write operations.
 /// It works best if all read-write-operation addresses are fixed offsets relative to some
 /// symbolic base address. If stack and heap access operations are mixed, this is usually violated.
-pub fn optimize_memory<T: FieldElement>(
-    mut machine: SymbolicMachine<T>,
+pub fn optimize_memory<T: FieldElement, V: Hash + Eq + Clone + Ord + Display>(
+    mut system: ConstraintSystem<T, V>,
     memory_bus_id: u64,
-) -> SymbolicMachine<T> {
+    range_constraints: impl RangeConstraintProvider<T, V> + Clone,
+) -> ConstraintSystem<T, V> {
     let (to_remove, new_constraints) =
-        redundant_memory_interactions_indices(&machine, memory_bus_id);
+        redundant_memory_interactions_indices(&system, memory_bus_id, range_constraints);
     let to_remove = to_remove.into_iter().collect::<HashSet<_>>();
-    machine.bus_interactions = machine
+    system.bus_interactions = system
         .bus_interactions
         .into_iter()
         .enumerate()
         .filter_map(|(i, bus)| (!to_remove.contains(&i)).then_some(bus))
         .collect();
-    machine.constraints.extend(new_constraints);
-    machine
+    // TODO perform substitutions instead
```

Comment:
Why? I think the inliner would do that, no? And maybe in some cases it wouldn't be possible (because of degree bound etc).